### PR TITLE
Add token type ids to CodeGenTokenizer

### LINKED ISSUE #28098: Create create_token_type_ids_from_sequences for CodeGenTokenizer
### Feature request

In CodeGenTokenizer [here](src/transformers/models/codegen/tokenization_codegen.py), there is no implementation for create_token_type_ids_from_sequences.

I was looking to the tutorial for token_type_ids as a reference: [here](https://huggingface.co/docs/transformers/glossary#token-type-ids). 

### Motivation

The model in input can require the token_type_ids [see here](https://huggingface.co/docs/transformers/model_doc/codegen#transformers.CodeGenForCausalLM), so would be useful to add this feature.

### Your contribution

I think it can be used directly the one frome codebert [here](src/transformers/models/bert/tokenization_bert.py):

````
    def create_token_type_ids_from_sequences(
        self, token_ids_0: List[int], token_ids_1: Optional[List[int]] = None
    ) -> List[int]:
        """
        Create a mask from the two sequences passed to be used in a sequence-pair classification task. A BERT sequence
        pair mask has the following format:

        ```
        0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1
        | first sequence    | second sequence |
        ```

        If `token_ids_1` is `None`, this method only returns the first portion of the mask (0s).

        Args:
            token_ids_0 (`List[int]`):
                List of IDs.
            token_ids_1 (`List[int]`, *optional*):
                Optional second list of IDs for sequence pairs.

        Returns:
            `List[int]`: List of [token type IDs](../glossary#token-type-ids) according to the given sequence(s).
        """
        sep = [self.sep_token_id]
        cls = [self.cls_token_id]
        if token_ids_1 is None:
            return len(cls + token_ids_0 + sep) * [0]
        return len(cls + token_ids_0 + sep) * [0] + len(token_ids_1 + sep) * [1]
````

### PR DESCRIPTION
**Pull request by st81**
# What does this PR do?

Fixes #28098 

## Who can review?

@ArthurZucker @younesbelkada

### CODE REVIEW

**Comment by st81** [comment]
This PR is WIP because I will check documentation update is needed and CodeGenTokenizer output includes `token_type_ids`.

**Comment by st81** [comment]
@ArthurZucker @younesbelkada
I've made the necessary changes to address the issue. Could someone please review the updated pull request at your earliest convenience? Your feedback is greatly appreciated.

Thank you!

**Review by ArthurZucker — COMMENTED** [review]
Thanks! You might have to update the `test_tokenizer_integration` as it will now return token_type_ids. Let's make sure slow tokenizer tests pass as well!

**Comment by st81** [comment]
> Thanks! You might have to update the `test_tokenizer_integration` as it will now return token_type_ids. Let's make sure slow tokenizer tests pass as well!

@ArthurZucker 

Thank you for taking the time to review the pull request! It seems that `test_tokenizer_integration` is missing from `tests/models/codegen/test_tokenization_codegen.py`. Would you like me to include `test_tokenizer_integration` similar to the other tokenizers tests?

**Comment by st81** [comment]
> Thanks! You might have to update the `test_tokenizer_integration` as it will now return token_type_ids. Let's make sure slow tokenizer tests pass as well!

@ArthurZucker 
It seems that `test_tokenizer_integration` is missing from `tests/models/codegen/test_tokenization_codegen.py`. So I’ve added `test_tokenizer_integration`. Could you review again?

**Comment by ArthurZucker** [comment]
Will review tomorrow sorry !

File: src/transformers/models/codegen/tokenization_codegen_fast.py [hunk 1]
```diff
@@ -177,6 +176,35 @@ def _encode_plus(self, *args, **kwargs) -> BatchEncoding:
 
         return super()._encode_plus(*args, **kwargs)
 
+    def create_token_type_ids_from_sequences(
```
Comment (ArthurZucker) [inline src/transformers/models/codegen/tokenization_codegen_fast.py:179]: let's use copied from here! (from the slow code)

**Review by ArthurZucker — COMMENTED** [review]
LGTM just a small nit

File: src/transformers/models/codegen/tokenization_codegen_fast.py (hunk 1 above)
Comment (st81) [reply src/transformers/models/codegen/tokenization_codegen_fast.py:180]: I have added `Copied from ...` comment!
(Please let me know if this change is not what you intended)

**Comment by st81** [comment]
I've addressed your comment and also added documentation for `CodeGenTokenizer.create_token_type_ids_from_sequences` ! Could you please review it again?

File: src/transformers/models/codegen/tokenization_codegen_fast.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/codegen/tokenization_codegen_fast.py:180]: yes! 🤗

File: tests/models/codegen/test_tokenization_codegen.py [hunk 2]
```diff
@@ -263,3 +263,10 @@ def test_truncation(self):
...
+        # fmt: on
+        self.tokenizer_integration_test_util(expected_encoding, "Salesforce/codegen-350M-mono", padding=False)
```
Comment (ArthurZucker) [inline tests/models/codegen/test_tokenization_codegen.py:272]:
```suggestion
        self.tokenizer_integration_test_util(expected_encoding, "Salesforce/codegen-350M-mono", padding=False, return_token_type_ids=True)
```
for BC we should make sure that `return_token_type_ids` is set to False by default

**Review by ArthurZucker — COMMENTED** [review]
Currently with the state of this PR:

```python 
In [1]:  from transformers import AutoTokenizer

In [2]: tokenizer = AutoTokenizer.from_pretrained("Salesforce/codegen-350M-mono")

In [3]:  tokenizer("Hey how are you")
Out[3]: {'input_ids': [10814, 703, 389, 345], 'token_type_ids': [0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1]}

```
this is not backward compatible. `return_token_type_ids` should be set to false

File: tests/models/codegen/test_tokenization_codegen.py (hunk 2 above)
Comment (st81) [reply tests/models/codegen/test_tokenization_codegen.py]: I've updated this test as you suggested.

File: tests/models/codegen/test_tokenization_codegen.py [hunk 3]
```diff
@@ -263,3 +263,12 @@ def test_truncation(self):
     # tokenizer has no padding token
     def test_padding_different_model_input_name(self):
         pass
+
+    # @slow
```
Comment (ArthurZucker) [inline tests/models/codegen/test_tokenization_codegen.py:267]:
```suggestion
    @slow
```

**Review by ArthurZucker — COMMENTED** [review]
We have to change the default `return_token_type_ids` not in the test (it's alright to keep it) but most importantly in the __init__ of both tokenizers, set it to `        self.return_token_type_ids = return_token_type_ids` and `return_token_type_ids=False` in the args!

File: tests/models/codegen/test_tokenization_codegen.py (hunk 3 above)
Comment (st81) [reply tests/models/codegen/test_tokenization_codegen.py]: Sorry I forgot to uncomment it. I've fixed it!

**Comment by st81** [comment]
@ArthurZucker 

Could you take a look at [this comment](https://github.com/huggingface/transformers/pull/29265#issuecomment-1983580444)? Your feedback is greatly appreciated.

**Comment by st81** [comment]
> @st81 Just to let you know @ArthurZucker is off for a week

Thank you for letting me know about that! I'll wait for him to return.

**Comment by st81** [comment]
@ArthurZucker 

> We have to change the default `return_token_type_ids` not in the test (it's alright to keep it) but most importantly in the **init** of both tokenizers, set it to ` self.return_token_type_ids = return_token_type_ids` and `return_token_type_ids=False` in the args!

I think current changes are fine to make the default `return_token_type_ids` is `False` because [it looks like the default value of `return_token_type_ids` is controlled by whether `self.model_input_names` contains `"token_type_ids"` or not](https://github.com/huggingface/transformers/blob/8dd4ce6f2cea1316cb5bd5ccbd348310ced17856/src/transformers/tokenization_utils_base.py#L3494-L3495). What are your thoughts on this?

**Review by ArthurZucker — COMMENTED** [review]
Hey, regarding: 
> When looking at other tokenizers in the repository, I found that none of them have return_token_type_ids as an argument in the __init__ function, and set return_token_type_ids as a class variable of the tokenizer like self.return_token_type_ids = return_token_type_ids. Would it be problematic from the perspective of library consistency to make CodeGenTokenizer the only one with such a specification?

it is not problematic no, as this is just a means to make sure we respect backward compatibility. 

Let's set it to False for this model only.

**Comment by st81** [comment]
> We have to change the default `return_token_type_ids` not in the test (it's alright to keep it) but most importantly in the **init** of both tokenizers, set it to ` self.return_token_type_ids = return_token_type_ids` and `return_token_type_ids=False` in the args!

I've made changes based on your [comment](https://github.com/huggingface/transformers/pull/29265#pullrequestreview-1921451566). Could you take a look?

File: src/transformers/models/codegen/tokenization_codegen.py [hunk 4]
```diff
@@ -150,13 +152,17 @@ def __init__(
...
         **kwargs,
     ):
         bos_token = AddedToken(bos_token, special=True) if isinstance(bos_token, str) else bos_token
         eos_token = AddedToken(eos_token, special=True) if isinstance(eos_token, str) else eos_token
         unk_token = AddedToken(unk_token, special=True) if isinstance(unk_token, str) else unk_token
         pad_token = AddedToken(pad_token, special=True) if isinstance(pad_token, str) else pad_token
         self.add_bos_token = add_bos_token
+        self.return_token_type_ids = return_token_type_ids
+        if self.return_token_type_ids:
+            self.model_input_names.append("token_type_ids")
```
Comment (ArthurZucker) [inline src/transformers/models/codegen/tokenization_codegen.py:165]: if we always add it does it make a difference? It's my first time seeing this fix, so a bit suprised that we have to do that

File: src/transformers/models/codegen/tokenization_codegen.py [hunk 5]
```diff
@@ -150,13 +152,17 @@ def __init__(
         pad_token=None,
         add_prefix_space=False,
         add_bos_token=False,
+        return_token_type_ids=False,
```
Comment (ArthurZucker) [inline src/transformers/models/codegen/tokenization_codegen.py:155]: that is the expected change

**Review by ArthurZucker — COMMENTED** [review]
LGTM otherwise

File: src/transformers/models/codegen/tokenization_codegen.py (hunk 4 above)
Comment (st81) [reply src/transformers/models/codegen/tokenization_codegen.py:165]: Yes. This actually controls whether the tokenizer returns `token_type_ids` or not in [here](https://github.com/huggingface/transformers/blob/8dd4ce6f2cea1316cb5bd5ccbd348310ced17856/src/transformers/tokenization_utils_base.py#L3494-L3495). The tokenizer doesn't return `token_type_ids` without this line.

Comment (ArthurZucker) [reply src/transformers/models/codegen/tokenization_codegen.py:165]: Sorry you are right! It's not really well made. I'll approve, but would be nice if we can fix self.model_iinput_names based on the value of `return_token_type_ids`. For now only the reverse is True.

**Review by ArthurZucker — APPROVED** [review]
Thanks

**Comment by ArthurZucker** [comment]
Sorry for the delay, LGTM

**Comment by st81** [comment]
Thank you so much for reviewing!

**Pull request merged**